Framework/Crystal/src/SCDCalibratePanels.cpp

```diff
@@ -605,677 +152,374 @@ void SCDCalibratePanels::exec() {
-      iFunc->setAttributeValue("BankNames", BankNameString);
-      iFunc->setAttributeValue("startX", -1);
-      iFunc->setAttributeValue("endX", -1);
-      iFunc->setAttributeValue("RotateCenters", RotGroups);
-      iFunc->setAttributeValue("SampleOffsets", SampOffsets);
-      iFunc->setParameter("l0", L0_bank);
-      iFunc->setParameter("t0", T0_bank);
-
-      double maxXYOffset = getProperty("MaxPositionChange_meters");
-
-      boost::shared_ptr<const RectangularDetector> bank_rect;
-
-      string name = group->front();
-      boost::shared_ptr<const IComponent> bank_cmp =
-          instrument->getComponentByName(name);
-      bank_rect =
-          boost::dynamic_pointer_cast<const RectangularDetector>(bank_cmp);
-
-      if (!bank_rect) {
-        g_log.error("No Rectangular detector bank " + banksVec[0] +
-                    " in instrument");
-        throw invalid_argument("No Rectangular detector bank " + banksVec[0] +
-                               " in instrument");
-      }
+  int nPeaks = static_cast<int>(peaksWs->getNumberPeaks());
+  bool changeL1 = getProperty("ChangeL1");
+  bool changeSize = getProperty("ChangePanelSize");
+  std::vector<std::string> fit_workspaces;
+  std::vector<std::string> parameter_workspaces;
+
+  if (changeL1)
+    findL1(nPeaks, peaksWs);
+  set<string> MyBankNames;
+  for (int i = 0; i < nPeaks; ++i) {
+    MyBankNames.insert(peaksWs->getPeak(i).getBankName());
+  }
 
-      // if( it1 == (*itv).begin())
-      CalcInitParams(bank_rect, instrument, PreCalibinstrument, detWidthScale0,
-                     detHeightScale0, Xoffset0, Yoffset0, Zoffset0, Xrot0,
-                     Yrot0, Zrot0);
-
-      // --- set Function property ----------------------
-      iFunc->setParameter("f0_detWidthScale", detWidthScale0);
-      iFunc->setParameter("f0_detHeightScale", detHeightScale0);
-      iFunc->setParameter("f0_Xoffset", Xoffset0);
-      iFunc->setParameter("f0_Yoffset", Yoffset0);
-      iFunc->setParameter("f0_Zoffset", Zoffset0);
-      iFunc->setParameter("f0_Xrot", Xrot0);
-      iFunc->setParameter("f0_Yrot", Yrot0);
-      iFunc->setParameter("f0_Zrot", Zrot0);
-
-      set<string> MyBankNames;
-      for (int i = 0; i < nPeaks; ++i)
-        MyBankNames.insert(banksVec[0]);
-      int startX = bounds[0];
-      int endXp1 = bounds[group->size()];
-      if (endXp1 - startX < 13) {
-        g_log.error() << "Bank Group " << BankNameString
-                      << " does not have enough peaks for fitting\n";
-        saveIsawDetCal(instrument, MyBankNames, T0_bank,
-                       DetCalFileName + std::to_string(iGr));
-        continue;
+  PARALLEL_FOR1(peaksWs)
+  for (int i = 0; i < static_cast<int>(MyBankNames.size()); ++i) {
+    PARALLEL_START_INTERUPT_REGION
+    std::set<string>::iterator it = MyBankNames.begin();
+    advance(it, i);
+    std::string iBank = *it;
+    const std::string bankName = "__PWS_" + iBank;
```

Comment:
Calling `advance` on a bidirectional iterator is O(n) and may make this loop slower than its serial equivalent. What does your profiling data show? Would `boost::container::flat_set` with random access iterators be a better choice?